File: src/bindings/python/src/pyopenvino/frontend/paddle/frontend.cpp
// Copyright (C) 2018-2021 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
//

#include "frontend.hpp"

#include <pybind11/functional.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/stl_bind.h>

#include "openvino/frontend/paddle/extension/conversion.hpp"
#include "openvino/frontend/paddle/frontend.hpp"
#include "openvino/frontend/paddle/node_context.hpp"

namespace py = pybind11;

using namespace ov::frontend::paddle;

#define CHECK_RET(any, T)                   \
    {                                       \
        if ((any).is<T>())                  \
            return py::cast((any).as<T>()); \
    }

void regclass_frontend_paddle_FrontEnd(py::module m) {
    py::class_<FrontEnd, std::shared_ptr<FrontEnd>> fem(m, "FrontEndPaddle", py::dynamic_attr(), py::module_local());
    fem.doc() = "openvino.frontend.paddle.FrontEnd wraps ov::frontend::paddle::FrontEnd";

    fem.def(py::init([]() {
        return std::make_shared<FrontEnd>();
    }));

    fem.def("convert",
            static_cast<std::shared_ptr<ov::Model> (FrontEnd::*)(const ov::frontend::InputModel::Ptr&) const>(
                &FrontEnd::convert),
            py::arg("model"));

    fem.def("convert",
            static_cast<void (FrontEnd::*)(const std::shared_ptr<ov::Model>&) const>(&FrontEnd::convert),
            py::arg("function"));

    fem.def("decode", &FrontEnd::decode, py::arg("model"));

    fem.def("get_name", &FrontEnd::get_name);

    fem.def("add_extension",
            static_cast<void (FrontEnd::*)(const std::shared_ptr<ov::Extension>& extension)>(&FrontEnd::add_extension));

    fem.def("__repr__", [](const FrontEnd& self) -> std::string {
        return "<FrontEnd '" + self.get_name() + "'>";
    });

    fem.def("convert_partially", &FrontEnd::convert_partially, py::arg("model"));

    fem.def("decode", &FrontEnd::decode, py::arg("model"));
}

void regclass_frontend_paddle_ConversionExtension(py::module m) {
    py::class_<ConversionExtension, ConversionExtension::Ptr, ov::frontend::ConversionExtensionBase> ext(
        m,
        "ConversionExtensionPaddle",
        py::dynamic_attr());

    ext.def(py::init([](const std::string& op_type, const ov::frontend::CreatorFunctionNamed& f) {
        return std::make_shared<ConversionExtension>(op_type, f);
    }));
}

File: src/bindings/python/src/openvino/frontend/frontend.py
# -*- coding: utf-8 -*-
# Copyright (C) 2018-2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

from typing import Union, Optional, cast, TYPE_CHECKING

from openvino._pyopenvino import FrontEnd as FrontEndBase
from openvino._pyopenvino import FrontEndManager as FrontEndManagerBase
from openvino._pyopenvino import InputModel
from openvino import Model


class FrontEnd(FrontEndBase):
    def __init__(self, fe: FrontEndBase) -> None:
        super().__init__(fe)

    def convert(self, model: Union[Model, InputModel]) -> Model:
        converted_model = super().convert(model)
        if isinstance(model, InputModel):
            return Model(converted_model)  # type: ignore
        return cast(Model, converted_model)

    def convert_partially(self, model: InputModel) -> Model:
        return Model(super().convert_partially(model))  # type: ignore

    def decode(self, model: InputModel) -> Model:
        return Model(super().decode(model))  # type: ignore

    def normalize(self, model: Model) -> None:
        super().normalize(model)


class FrontEndManager(FrontEndManagerBase):
    def load_by_framework(self, framework: str) -> Optional[FrontEnd]:
        fe = super().load_by_framework(framework)
        if fe is not None:
            return FrontEnd(fe)
        return None

    def load_by_model(self, model: str) -> Optional[FrontEnd]:
        fe = super().load_by_model(model)
        if fe is not None:
            return FrontEnd(fe)
        return None